### ISSUE TITLE
Feedback from Facebook

### ISSUE THREAD

**Issue by vdjeric** [WICG/performance-memory#7]
This is feedback from Benoit and Vlad, Nate will reply separately.

We really like this proposal: private memory footprint makes a lot of sense, "accurate or null" is a huge improvement, and it seems like most privacy issues have been addressed.

## Things that are unclear to us:

> For privateMemoryFootprint, totalJSHeapSize, and usedJSHeapSize we only have accurate accounting when the process is hosting a single top level frame. As such, we return null anytime the process is hosting more than a single top level frame.

What will be returned when two Facebook.com tabs are in the same render process? We would prefer null for this situation, since it is very hard to interpret a single memory usage number from multiple tabs.

> One strange thing in this proposal is that web authors will need to handle that some instances of Chrome (e.g. on mobile) won’t have full site isolation and others (e.g. on desktop) will. It’s a bit ugly, but they can detect this by looking at cases where all the values in that page ended in four zeros. If all calls to performance.memory result in quantized values, then it’s almost guaranteed to be the quantized numbers.

This is from the proposal in the Google Doc https://docs.google.com/document/d/1iC9-sra-vdz0c18qM9H_El_k4yGzd6TdIItIfs3p-8Q/edit#

So does this mean that mobile Chrome will ALWAYS quantize and delay memory measurements?
What if there is a single tab in mobile Chrome, will it return the accurate measurement?
In the past, we have used the quantized/delayed performance.memory API and have not gotten benefit out of it. So we would prefer "accurate or null" again -- either the measurement is from a single tab or it is null.

> The results are non-null if and only if all frames hosted by the process are part of a single frame tree.

Can you clarify, perhaps in the explainer, exactly what this means for the following scenarios:
- Page with a same-origin iframe
- Page with a foreign-origin iframe but OOPIF enabled

## Our suggestions:

- The spec should have a provision to allow the browser to return NULL at its discretion to protect the user's privacy. So for example, the browser might want to return null from the memory API if a foreign origin IFRAME was every loaded by the page (i.e. if the renderer process was tainted with foreign-origin data).
Example attack: A malicious top-level frame measures its memory usage, then loads an IFRAME pointed to a target origin, then destroys this IFRAME and measures memory again, relying on the browser to not have freed all memory from the foreign origin immediately. This would allow the malicious top-level frame to obtain information about the target page, e.g. whether the user was logged into the target origin.
- Is it possible to get additional categories in the future to make the memory information more actionable? Examples: memory used by HTML nodes, CSS, Canvas, audio/video elements, etc. For instance, we have had situations where React was leaking nodes, and this type of information could have made it easier to find the cause.

**Comment by erikchen** [comment]
Thanks for the detailed feedback!

> "accurate or null" is a huge improvement, and it seems like most privacy issues have been addressed.

After [discussions](https://github.com/erikchen/performance-memory/issues/8) with Gsuite & gmail, we decided to change the API to:
* include numberOfTabs
* return non-null numbers, even when numberOfTabs > 1

The recommendation for most developers is to ignore the results when numberOfTabs > 1. We probably still need to work on ergonomics of the API, but punting that for later as a superficial change.

> What will be returned when two Facebook.com tabs are in the same render process? 

Added a more detailed examples section in Appendix D. Under the new spec, we will return numberOfTabs = 2 [along with non-null memory numbers]

> So does this mean that mobile Chrome will ALWAYS quantize and delay memory measurements?

This API assumes site-isolation is enabled. The proposal is to return |null| if the conditions cannot be met. This means that for some percentage of mobile devices, we will return |null|. Not sure what that percentage is. 

> Can you clarify, perhaps in the explainer, exactly what this means for the following scenarios:

Done, see Appendix D.

> The spec should have a provision to allow the browser to return NULL at its discretion to protect the user's privacy.
We added the following change:
```
// Any or all of these values may be |null| if:
//  - The browser is unable to provide accurate information due to
//    implementation limitations.
//  - Exposing the information could leak information across frames from
//    different origins.
```

> Is it possible to get additional categories in the future to make the memory information more actionable?
We would like to do this, but this requires coming up with consistent definitions. Added this request as Appendix E

**Comment by n8schloss** [comment]
How do service workers / other web workers fit into the numberOfTabs metric? Perhaps there could be a numberOfWorkers metric that shows how many workers are in the process?

My thinking here is that sometimes the sw might be in a different process from a tab even if numberOfTabs = 1, but sometimes it might be in the same process. So it'd be useful to be able to tell that apart too.

**Comment by erikchen** [comment]
Good point. We added a new field: numberOfWorkers in this [pull request](https://github.com/erikchen/performance-memory/pull/11/files)
